File: src/readers/gbfs/schemaV1/gbfs.ts
/**
 * # GBFS Schema V1.1 OR GBFS Schema V1.0
 * Auto-discovery file that links to all of the other files published by the system.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#gbfsjson)
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#gbfsjson)
 */
export type GBFSV1 = GBFSV11 | GBFSV10;

/**
 * # GBFS Schema V1.1
 * Auto-discovery file that links to all of the other files published by the system.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#gbfsjson)
 */
export const gbfsSchemaV11 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#gbfsjson',
  description: 'Auto-discovery file that links to all of the other files published by the system.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 1450155600,
    },
    ttl: {
      description:
        'Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).',
      type: 'integer',
      minimum: 0,
    },
    version: {
      description:
        'GBFS version number to which the feed conforms, according to the versioning framework (added in v1.1).',
      type: 'string',
      const: '1.1',
    },
    data: {
      description: 'Response data in the form of name:value pairs.',
      type: 'object',
      patternProperties: {
        '^[a-z]{2,3}(-[A-Z]{2})?$': {
          type: 'object',
          properties: {
            feeds: {
              description:
                'An array of all of the feeds that are published by the auto-discovery file.',
              type: 'array',
              items: {
                type: 'object',
                properties: {
                  name: {
                    description: 'Key identifying the type of feed.',
                    type: 'string',
                    enum: [
                      'gbfs',
                      'gbfs_versions',
                      'system_information',
                      'station_information',
                      'station_status',
                      'free_bike_status',
                      'system_hours',
                      'system_alerts',
                      'system_calendar',
                      'system_regions',
                      'system_pricing_plans',
                    ],
                  },
                  url: {
                    description: 'URL for the feed.',
                    type: 'string',
                    format: 'uri',
                  },
                },
                required: ['name', 'url'],
              },
            },
          },
          required: ['feeds'],
        },
      },
      minProperties: 1,
      additionalProperties: false,
    },
  },
  required: ['last_updated', 'ttl', 'version', 'data'],
};

/**
 * GBFS Schema V1.1 Interface
 */
export interface GBFSV11 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;

  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;

  /** GBFS version number (1.1). */
  version: '1.1';

  /** Response data in the form of name:value pairs. */
  data: {
    [locale: string]: {
      feeds: Array<{
        name:
          | 'gbfs'
          | 'gbfs_versions'
          | 'system_information'
          | 'station_information'
          | 'station_status'
          | 'free_bike_status'
          | 'system_hours'
          | 'system_alerts'
          | 'system_calendar'
          | 'system_regions'
          | 'system_pricing_plans';
        url: string;
      }>;
    };
  };
}

/**
 * # GBFS Schema V1.0
 * Auto-discovery file that links to all of the other files published by the system.
 *
 * ## Links
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#gbfsjson)
 */
export const gbfsSchemaV10 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#gbfsjson',
  description: 'Auto-discovery file that links to all of the other files published by the system.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 0,
      maximum: 1924988399,
    },
    ttl: {
      description: 'Number of seconds before the data in the feed will be updated again.',
      type: 'integer',
      minimum: 0,
    },
    data: {
      description: 'Response data in the form of name:value pairs.',
      type: 'object',
      patternProperties: {
        '^[a-zA-Z]{2}$': {
          type: 'object',
          properties: {
            feeds: {
              description:
                'An array of all of the feeds that are published by the auto-discovery file.',
              type: 'array',
              items: {
                type: 'object',
                properties: {
                  name: { type: 'string' },
                  url: { type: 'string' },
                },
                required: ['name', 'url'],
              },
            },
          },
          required: ['feeds'],
        },
      },
      minProperties: 1,
      additionalProperties: false,
    },
  },
  required: ['last_updated', 'ttl', 'data'],
};

/**
 * GBFS Schema V1.0 Interface
 */
export interface GBFSV10 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;

  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;

  /** Response data in the form of name:value pairs. */
  data: {
    [locale: string]: {
      feeds: Array<{
        name: string;
        url: string;
      }>;
    };
  };
}

File: rust/readers/gbfs/schema_v1/gbfs.rs
use alloc::{collections::BTreeMap, string::String, vec::Vec};
use serde::{Deserialize, Serialize};

/// # GBFS Schema V1.1 OR GBFS Schema V1.0
/// Auto-discovery file that links to all of the other files published by the system.
///
/// ## Links
/// - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#gbfsjson)
/// - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#gbfsjson)
pub type GBFSV1 = GBFSV11;

/// GBFS Schema V1.1 Feeds Names
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub enum GBFSV11FeedsName {
    /// GBFS
    #[default]
    #[serde(rename = "gbfs")]
    Gbfs,
    /// GBFS Versions
    #[serde(rename = "gbfs_versions")]
    GbfsVersions,
    /// System Information
    #[serde(rename = "system_information")]
    SystemInformation,
    /// Station Information
    #[serde(rename = "station_information")]
    StationInformation,
    /// Station Status
    #[serde(rename = "station_status")]
    StationStatus,
    /// Free Bike Status
    #[serde(rename = "free_bike_status")]
    FreeBikeStatus,
    /// System Hours
    #[serde(rename = "system_hours")]
    SystemHours,
    /// System Alerts
    #[serde(rename = "system_alerts")]
    SystemAlerts,
    /// System Calendar
    #[serde(rename = "system_calendar")]
    SystemCalendar,
    /// System Regions
    #[serde(rename = "system_regions")]
    SystemRegions,
    /// System Pricing Plans
    #[serde(rename = "system_pricing_plans")]
    SystemPricingPlans,
}

/// GBFS Schema V1.1 Feeds
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV11Feeds {
    /// Name of the feed
    pub name: GBFSV11FeedsName,
    /// URL of the feed
    pub url: String,
}

/// GBFS Schema V1.1 Feeds Data
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV11FeedsData {
    /// List of feeds
    pub feeds: Vec<GBFSV11Feeds>,
}

/// GBFS Schema V1.1 Interface
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV11 {
    /// Last time the data in the feed was updated in POSIX time.
    pub last_updated: u64,
    /// Number of seconds before the data in the feed will be updated again.
    pub ttl: u64,
    /// GBFS version number (1.1).
    pub version: String,
    /// Response data in the form of name:value pairs.
    pub data: BTreeMap<String, GBFSV11FeedsData>,
}

/// GBFS Schema V1.0 Feeds
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV10Feeds {
    /// Name of the feed
    pub name: String,
    /// URL of the feed
    pub url: String,
}

/// GBFS Schema V1.0 Feeds Data
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV10FeedsData {
    /// List of feeds
    feeds: Vec<GBFSV10Feeds>,
}

/// GBFS Schema V1.0 Interface
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSV10 {
    /// Last time the data in the feed was updated in POSIX time.
    pub last_updated: u64,
    /// Number of seconds before the data in the feed will be updated again.
    pub ttl: u64,
    /// Response data in the form of name:value pairs.
    pub data: BTreeMap<String, GBFSV10FeedsData>,
}
